### ISSUE TITLE
Download tracker does not track some chapters properly (due to special characters?)

### ISSUE THREAD

**Issue by CaptainLingonberry** [mansuf/mangadex-downloader#51]
### What happened ?

The downloaded-cbz.json doesn't seem to track certain chapters of this manga correctly: https://mangadex.org/title/9d3d3403-1a87-4737-9803-bc3d99db1424. This may be due to a special character used in some titles.

It will re-download chapters 20, 27, 28, 29, 34, and 37 every time the command is run, even if all chapters have been downloaded before. I've noticed that these particular chapters have a [right single quotation mark](https://en.wikipedia.org/wiki/Right_single_quotation_mark) in the titles rather than straight apostrophes, and have garbled titles in the downloaded-cbz.json, like "SheÃƒÂ¢Ã¢â€šÂ¬Ã¢â€žÂ¢s the Only One I Refuse to Share.cbz". The cbz will still be downloaded with the correct title.

If this is the cause, there may also be problems with other special characters in chapter titles.

This is the command's output:

```
[INFO] Checking url = https://mangadex.org/title/9d3d3403-1a87-4737-9803-bc3d99db1424
[INFO] Using English language
[INFO] Fetching all chapters...
[INFO] Downloading cover manga The Guy She Was Interested in Wasn't a Guy at All
[INFO] Not downloading cover manga, since "cover" is none
[INFO] Download directory is set to "C:\Users\User\mangadex-dl\manga\The Guy She Was Interested in Wasn't a Guy at All"
[INFO] Using cbz format
[INFO] Preparing to download...
[INFO] Getting images from chapter 20
[INFO] Downloading Chapter. 20 - When It Comes to His Niece, There’s No Spending Too Much page 1
file_sizes: 100%|███████████████████████████| 2.22M/2.22M [00:04<00:00, 536kB/s]
[INFO] Downloading Chapter. 20 - When It Comes to His Niece, There’s No Spending Too Much page 2
file_sizes: 100%|███████████████████████████| 2.50M/2.50M [00:03<00:00, 682kB/s]
[INFO] Downloading Chapter. 20 - When It Comes to His Niece, There’s No Spending Too Much page 3
file_sizes: 100%|███████████████████████████| 2.84M/2.84M [00:04<00:00, 657kB/s]
[INFO] Downloading Chapter. 20 - When It Comes to His Niece, There’s No Spending Too Much page 4
file_sizes: 100%|███████████████████████████| 2.36M/2.36M [00:07<00:00, 307kB/s]
[INFO] Ch. 20 - When It Comes to His Niece, There’s No Spending Too Much has finished download, converting to cbz...
cbz_progress: 100%|████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 95.30item/s]
[INFO] Getting images from chapter 27
[INFO] Downloading Chapter. 27 - I’m the Only One Who Knows Who My Crush (A Girl) Is page 1
file_sizes: 100%|███████████████████████████| 2.66M/2.66M [00:03<00:00, 836kB/s]
[INFO] Downloading Chapter. 27 - I’m the Only One Who Knows Who My Crush (A Girl) Is page 2
file_sizes: 100%|███████████████████████████| 2.32M/2.32M [00:02<00:00, 785kB/s]
[INFO] Downloading Chapter. 27 - I’m the Only One Who Knows Who My Crush (A Girl) Is page 3
file_sizes: 100%|███████████████████████████| 2.74M/2.74M [00:03<00:00, 761kB/s]
[INFO] Downloading Chapter. 27 - I’m the Only One Who Knows Who My Crush (A Girl) Is page 4
file_sizes: 100%|███████████████████████████| 2.21M/2.21M [00:04<00:00, 538kB/s]
[INFO] Ch. 27 - I’m the Only One Who Knows Who My Crush (A Girl) Is has finished download, converting to cbz...
cbz_progress: 100%|███████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 105.32item/s]
[INFO] Getting images from chapter 28
[INFO] Downloading Chapter. 28 - Someone’s Going to Get Blinded page 1
file_sizes: 100%|███████████████████████████| 2.30M/2.30M [00:18<00:00, 126kB/s]
[INFO] Downloading Chapter. 28 - Someone’s Going to Get Blinded page 2
file_sizes: 100%|███████████████████████████| 2.46M/2.46M [00:13<00:00, 182kB/s]
[INFO] Downloading Chapter. 28 - Someone’s Going to Get Blinded page 3
file_sizes: 100%|███████████████████████████| 3.03M/3.03M [00:18<00:00, 160kB/s]
[INFO] Downloading Chapter. 28 - Someone’s Going to Get Blinded page 4
file_sizes: 100%|███████████████████████████| 3.04M/3.04M [00:10<00:00, 294kB/s]
[INFO] Ch. 28 - Someone’s Going to Get Blinded has finished download, converting to cbz...
cbz_progress: 100%|████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 88.94item/s]
[INFO] Getting images from chapter 29
[INFO] Downloading Chapter. 29 - She’s the Only One I Refuse to Share page 1
file_sizes: 100%|███████████████████████████| 2.67M/2.67M [00:03<00:00, 760kB/s]
[INFO] Downloading Chapter. 29 - She’s the Only One I Refuse to Share page 2
file_sizes: 100%|███████████████████████████| 2.12M/2.12M [00:02<00:00, 733kB/s]
[INFO] Downloading Chapter. 29 - She’s the Only One I Refuse to Share page 3
file_sizes: 100%|███████████████████████████| 3.29M/3.29M [00:05<00:00, 562kB/s]
[INFO] Downloading Chapter. 29 - She’s the Only One I Refuse to Share page 4
file_sizes: 100%|███████████████████████████| 2.60M/2.60M [00:04<00:00, 625kB/s]
[INFO] Ch. 29 - She’s the Only One I Refuse to Share has finished download, converting to cbz...
cbz_progress: 100%|████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 90.97item/s]
[INFO] Getting images from chapter 34
[INFO] Downloading Chapter. 34 - Most Likely, It’s Not Her Imagination page 1
file_sizes: 100%|███████████████████████████| 2.64M/2.64M [00:14<00:00, 179kB/s]
[INFO] Downloading Chapter. 34 - Most Likely, It’s Not Her Imagination page 2
file_sizes: 100%|███████████████████████████| 2.70M/2.70M [00:06<00:00, 395kB/s]
[INFO] Downloading Chapter. 34 - Most Likely, It’s Not Her Imagination page 3
file_sizes: 100%|███████████████████████████| 3.03M/3.03M [00:05<00:00, 525kB/s]
[INFO] Downloading Chapter. 34 - Most Likely, It’s Not Her Imagination page 4
file_sizes: 100%|███████████████████████████| 2.61M/2.61M [00:09<00:00, 273kB/s]
[INFO] Ch. 34 - Most Likely, It’s Not Her Imagination has finished download, converting to cbz...
cbz_progress: 100%|████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 83.38item/s]
[INFO] Getting images from chapter 37
[INFO] Downloading Chapter. 37 - The World’s Most Beautiful Sound page 1
file_sizes: 100%|███████████████████████████| 2.85M/2.85M [00:09<00:00, 291kB/s]
[INFO] Downloading Chapter. 37 - The World’s Most Beautiful Sound page 2
file_sizes: 100%|███████████████████████████| 2.89M/2.89M [00:05<00:00, 487kB/s]
[INFO] Downloading Chapter. 37 - The World’s Most Beautiful Sound page 3
file_sizes: 100%|███████████████████████████| 2.82M/2.82M [00:04<00:00, 567kB/s]
[INFO] Downloading Chapter. 37 - The World’s Most Beautiful Sound page 4
file_sizes: 100%|███████████████████████████| 3.46M/3.46M [00:12<00:00, 282kB/s]
[INFO] Ch. 37 - The World’s Most Beautiful Sound has finished download, converting to cbz...
cbz_progress: 100%|████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 77.16item/s]
[INFO] 'Ch. 1.cbz' is verified and no need to re-download
[INFO] 'Ch. 2.cbz' is verified and no need to re-download
[INFO] 'Ch. 3.cbz' is verified and no need to re-download
[INFO] 'Ch. 4.cbz' is verified and no need to re-download
[INFO] 'Ch. 5.cbz' is verified and no need to re-download
[INFO] 'Ch. 6.cbz' is verified and no need to re-download
[INFO] 'Ch. 7.cbz' is verified and no need to re-download
[INFO] 'Ch. 8 - The Type of Girl Who Makes a Pass When She's Panicking.cbz' is verified and no need to re-download
[INFO] 'Ch. 9 - She Doesn't Have Friends, but She Felt Happy All the Same.cbz' is verified and no need to re-download
[INFO] 'Ch. 9.1 - Character Bios.cbz' is verified and no need to re-download
[INFO] 'Ch. 10.cbz' is verified and no need to re-download
[INFO] 'Ch. 10.1 - The Two of Them After Being Caught in the Rain.cbz' is verified and no need to re-download
[INFO] 'Ch. 11.cbz' is verified and no need to re-download
[INFO] 'Ch. 12.cbz' is verified and no need to re-download
[INFO] 'Ch. 13 - Girls With an Extreme Gap.cbz' is verified and no need to re-download
[INFO] 'Ch. 14 - Today Too, the Gyaru Is Involved With the Gloomy Girl.cbz' is verified and no need to re-download
[INFO] 'Ch. 15 - The Gloomy Girl in My Class Was Super Cool.cbz' is verified and no need to re-download
[INFO] 'Ch. 16 - The Voice of Her Crush (Actually a Girl) Is Unforgettable.cbz' is verified and no need to re-download
[INFO] 'Ch. 17 - Caught in The Storm.cbz' is verified and no need to re-download
[INFO] 'Ch. 18 - The Guy She Was Interested in Wasn't a Guy At All.cbz' is verified and no need to re-download
[INFO] 'Ch. 18.1 - The Two of Them Shopping on a Day Off, Five Years From Now (Hopefully).cbz' is verified and no need to re-download
[INFO] 'Ch. 19 - A New Semester of Don't Ask, Don't Tell.cbz' is verified and no need to re-download
[INFO] 'Ch. 20 - When It Comes to His Niece, There’s No Spending Too Much.cbz' is verified and no need to re-download
[INFO] 'Ch. 21 - Come On, Just Face Her.cbz' is verified and no need to re-download
[INFO] 'Ch. 22 - The Girl at the CD Store.cbz' is verified and no need to re-download
[INFO] 'Ch. 22.1 - Maybe It Would Be Like This if the Two of Them Moved in Together... Do Your Best!.cbz' is verified and no need to re-download
[INFO] 'Ch. 23 - Stop Right There!.cbz' is verified and no need to re-download
[INFO] 'Ch. 24 - I'll Keep Chasing You!.cbz' is verified and no need to re-download
[INFO] 'Ch. 25 - Feels Like a Dream.cbz' is verified and no need to re-download
[INFO] 'Ch. 26 - The Guy She Liked Was a Girl Friend.cbz' is verified and no need to re-download
[INFO] 'Ch. 26.1 - I Kind of Wanted to Make Some Goods That These Two Might Have.cbz' is verified and no need to re-download
[INFO] 'Ch. 27 - I’m the Only One Who Knows Who My Crush (A Girl) Is.cbz' is verified and no need to re-download
[INFO] 'Ch. 28 - Someone’s Going to Get Blinded.cbz' is verified and no need to re-download
[INFO] 'Ch. 29 - She’s the Only One I Refuse to Share.cbz' is verified and no need to re-download
[INFO] 'Ch. 30 - If It's These Two.cbz' is verified and no need to re-download
[INFO] 'Ch. 31 - An Unexpected Encounter With Her Crush (A Girl).cbz' is verified and no need to re-download
[INFO] 'Ch. 32 - Telepathy.cbz' is verified and no need to re-download
[INFO] 'Ch. 33 - Wish, and You Shall Receive.cbz' is verified and no need to re-download
[INFO] 'Ch. 34 - Most Likely, It’s Not Her Imagination.cbz' is verified and no need to re-download
[INFO] 'Ch. 35 - Peacemaker.cbz' is verified and no need to re-download
[INFO] 'Ch. 36 - Trouble Chaser.cbz' is verified and no need to re-download
[INFO] 'Ch. 37 - The World’s Most Beautiful Sound.cbz' is verified and no need to re-download
[INFO] 'Ch. 37.1 - Smoothly (Valentine's day twitter pic - Landscape and portrait orientation).cbz' is verified and no need to re-download
[INFO] Waiting for chapter read marker to finish
[INFO] Download finished for manga "The Guy She Was Interested in Wasn't a Guy at All"
[INFO] Cleaning up...
```

And this is the downloaded-cbz.json (renamed to txt so I could upload it):

[downloaded-cbz.txt](https://github.com/mansuf/mangadex-downloader/files/10775380/downloaded-cbz.txt)

### What did you expect to happen ?

It should properly verify the chapters are downloaded and not redownload anything.

### OS version

Windows 10 21H1

### App version

mangadex-downloader v2.7.1 (https://github.com/mansuf/mangadex-downloader)
Python: 3.10.9
arch: x64
bundled executable: yes

### Installation origin

Github releases

### Reproducible command

mangadex-dl "https://mangadex.org/title/9d3d3403-1a87-4737-9803-bc3d99db1424" --language en --use-chapter-title --no-group-name --cover none --save-as cbz --folder manga

### Additional context

The command is run from a batch file.

**Comment by CaptainLingonberry** [comment]
I forgot to test it before, but there is no problem if --use-chapter-title is omitted. The affected chapters are not re-downloaded on subsequent runs. However, I'd still prefer using chapter titles.

**Comment by mansuf** [comment]
Fixed in https://github.com/mansuf/mangadex-downloader/commit/1e9a06e96786bd7a6f7204880db4cce200fb9e7c, you can download patched application from [this link](https://github.com/mansuf/mangadex-downloader/actions/runs/4216388283) once it's show success result

The problem caused by string encoding. The `DownloadTracker._write` function is writing dumped JSON to a file with binary mode (`wb`), while `DownloadTracker._load` function is reading dumped JSON from a file with text mode (`r`), and thus string encoding error happened when loading JSON downloaded-cbz.json file

`DownloadTracker._load`
https://github.com/mansuf/mangadex-downloader/blob/0164b18b739ed743607db45443c62e58546b8993/mangadex_downloader/tracker.py#L407-L420

`DownloadTracker._write`
https://github.com/mansuf/mangadex-downloader/blob/0164b18b739ed743607db45443c62e58546b8993/mangadex_downloader/tracker.py#L247-L258

**Comment by mansuf** [comment]
Patch released in [v2.7.2](https://github.com/mansuf/mangadex-downloader/releases/tag/v2.7.2)
